README
Welcome to procinfo 18!

Procinfo is a small program that gathers some system information from
diverse files under /proc and prints it to the screen. It duplicates some of
the functionality of free(1) and uptime(1). It should work with any kernel
release close to 1.0 or later up to the latest development kernel. (Well, I
haven't actually tried 1.0 for years, but you get the idea.)

Lsdev is a Perl script that attempts to tabulate DMA, IRQ and I/O port
allocation per kernel driver.

Socklist is a Perl script that gives you a list of all open sockets,
enumerating types, port, inode, uid, pid, fd and the program to which
it belongs. Contributed by Larry Doolittle <ldoolitt@jlab.org>.

A list of changes for this and older versions of this package can be found
in the file CHANGES.

If you have any problems running procinfo under older kernels (from
1.0 onwards), don't hesitate to mail me about it. The code may be an
unsightly mess which should have been re-written from scratch ages ago
if I had thought it worth the effort, but at least we can pride
ourselves on staunchly sticking to backwards compatibility beyond your
wildest dreams.

Thanks to everyone who send bug reports/fixes, commentary, etc.


Have fun,

Sander van Malssen

svm@kozmix.cistron.nl -- http://www.cistron.nl/~svm/


CHANGES
* New in version 18 (2001-03-02):
=================================

Right, finally a new release of procinfo.

This release has a number of bugfixes, but I'm pretty sure I've lost
some and introduced some new ones along the way, for which my apologies.
This release is primarily to get a new baseline out for making patches
against, and to fix an annoying 2.4 incompatibility. I wouldn't start
sticking it into distributions just yet if I were you...

  + Disk stats are back again for 2.4 kernels. (Well, some of them. The
    kernel only keeps track of stats for block devices up to no. 16 or so,
    which for example excludes the second IDE controller, which sits at
    major number 22.)

  + Fix for Redhat Bugzilla bug 17391 (when uptime gets too high, procinfo
    displays 0.0% cpu related times) by Jon Lewis <jlewis@lewis.org>.

  + New feature: Use 'C' to temporarily zero out stats ('checkpointing'),
    and 'R' to release the checkpoint (only makes sense in totals mode).

  + Fixed a bug where procinfo would crash if run from a telnet client that
    doesn't support the TIOCWINSZ window-size ioctl (a.k.a. the Redmond
    Bug). (Thanks to Phil Winans <phil@jtsmith.com> for the bug report.)

  + New option -r adds a "-/+ buffers/cache:" line to the memory info bit,
    just like free(1) does.

  + Pressing space in full-screen mode now freezes the display until you
    press a key again.

  + Fixed a some non-Intel and/or SMP bugs (thanks to David Landgren
    <david@landgren.net>, michaelkjohnson <johnsonm@redhat.com>,Bernhard
    Rosenkraenzer <bero@redhat.de> and Chris Lumens <chris@slackware.com>).
    Folks, there's really no need for this kind of bugs to pass me by
    completely unnoticed if you all would just get together and donate me
    some 8-way Alphas or UltraSparcs, you know.


* New in version 17 (1999-05-16):
=================================

  + Sparc fixlet from RedHat (Jeff Johnson <jbj@redhat.com>). If you have a
    non-Intel-Alpha-Sparc Linux box and you know that timer interrupts are
    generated on some specific IRQ line, let me know.

  + Options -s (not that it is actually all that useful) went missing a
    while ago but is back again (David <djhill@ameritech.net>).

  + Internal cleanup of the irq stuff. (Should've done that ages ago, the
    code is noticeably cleaner now.) This also fixes a bug along the way
    where the context switch value would be clobbered over by the irq code.

  + Some minor tweaks to the procinfo manual page.

  + CPU percentages should work OK on SMP boxes now -- user + nice + system
    + idle now adds up to 100%, regardless of the number of CPU's (thanks to
    Chris Adams <cadams@ro.com> for reporting and testing). Since we had to
    start counting the number of CPUs anyway, procinfo now prints the number
    of CPU's on the top line.

  + Fixed lossage when you run procinfo full-screen in combination with -F
    or with redirection (thanks to Manfred Lemke <lema0019@fh-karlsruhe.de>
    for noting this).

  + Fixed display of modules, devices and filesystems in combination with
    odd screen sizes.


* New in version 16 (1998-12-31):
=================================

  + Hopefully made thing saner w.r.t. handling of large numbers of
    interrupts.

  + Fixed a problem with interrupts on SMP boxen (noted by dth). Note that
    procinfo will only display values for the first CPU.


* New in version 15 (1998-07-04):
=================================

  + Fixed IRQ bug introduced in version 14. Thanks to Samel Vitezslav
    <xsamel00@dcse.fee.vutbr.cz>, and Evgeni A. Nurminski
    <nurmi@ipm.marine.su>.

  + lsdev parses /proc/interrupts properly again. Since I fixed this by
    *simplifying* things I have the lingering suspicion I might have broken
    something somewhere else again, but it seems to work with 2.0.33 and
    2.1.103.  Thanks to Samel Vitezslav for letting me know.


* New in version 14 (1998-05-24):
=================================

  + Forgot to freopen stdin and stderr as well as stdout when using -F,
    leading to funny things in full-screen mode. Fixed.

  + Appearantly there are some SMP motherboard out there that can handle 24
    IRQ's instead of 16. Fixed, and also to make sure that things still fit
    inside 80x24, procinfo now squeezes the IRQ list by default. The -i
    option now unsqueezes the display. (Thank you Edwin Huffstutler
    <edwinh@computer.org>.)


* New in version 13 (1998-03-07):
=================================

  + Procinfo's gone interactive! :-) Pressing 'q' when running full-screen
    exits, pressing 'd' toggles between what the -d option does, etc. etc.
    etc. etc. etc.

  + Fixes for Linux/Alpha, relating to the obscene number of IRQ's on that
    platform. To make it possible to keep running procinfo full-screen on a
    80x24 terminal on Alphas, I've added an option -i that squeezes the IRQ
    portion of the display by only printing IRQ's that have either a label
    or a non-zero interrupt count. It's not guaranteed to work if you really
    *have* a lot of IRQ's allocated, of course, but you can't have
    everything.

  + Fixed a not-zero-terminated bug and a driver-name-starts-with-digit bug,
    both in the interrupts handling, both thanks to Johan Lindh
    <sirreal@warrior.net>.

  + Added a short manpage for socklist.

  + Signal handling cleaned up. This incidentally fixes a bug with
    suspending a procinfo compiled with glibc2, which I had brilliantly
    predicted way back when I wrote version 0.1 in 1994, long before I'd
    even *heard* of glibc2. Clever, huh?


* New in version 12 (1998-01-21):
=================================

  + New utility! Written by Larry Doolittle <ldoolitt@jlab.org>. Socklist
    gives you a list of all open sockets, enumerating types, port, inode,
    uid, pid, fd and the program to which it belongs. Thanks Larry!

  + With -d or -D and update intervals other than one second, the values

procinfo.h
/*

  procinfo.h

  Date:        1995-04-15 23:54:06
  Last Change: 2001-02-24 23:43:08

  Copyright (C) 1995-2001 Sander van Malssen <svm@kozmix.cistron.nl>

  This software is released under the GNU Public Licence. See the file
  `COPYING' for details. Since you're probably running Linux I'm sure
  your hard disk is already infested with copies of this file, but if
  not, mail me and I'll send you one.

  $Id: procinfo.h,v 1.25 2001/02/24 23:30:35 svm Exp svm $

*/

#ifndef _PROCINFO_H
#define _PROCINFO_H

#define VERSION		"18 (2001-03-02)"

#ifndef PROC_DIR
#define PROC_DIR	"/proc/"	/* Needs the trailing slash. */
#endif

#define ISSTR(s)	(!strcmp(s, type))

#if 0
#define VAL		(strtoul(strtok(NULL, "\t "), NULL, 10))
#else
static inline unsigned long find_val(void)
{
    char *cp;

    cp = strtok(NULL, "\t ");
    if (!cp)
	return 0L;
    return (strtoul (cp, NULL, 10));
}
#define VAL	find_val()
#endif

#define DIFF(x)		(show_diff ? \
			 (unsigned long) (((new.x)-(old.x))/rate) : \
			  new.x)

#define bDIFF(x)	(show_diff ? \
			 (unsigned long) (((new.x)-(old.x))/rate) : \
		 	 (show_from_baseline ? \
			  new.x - base.x : new.x))

#define COPYINFO(to,from)	{ \
    unsigned long *foo; \
    foo = to.intr; \
    memcpy (&to, &from, sizeof (struct info)); \
    to.intr = foo; \
    memcpy (to.intr, from.intr, nr_irqs * sizeof (unsigned long)); \
}



#define IS_ODD(x)	(x & 1)

#undef  MAX
#define MAX(a,b)	((a) > (b) ? (a) : (b))

#define CDRV		0
#define BDRV		1
#ifndef MAX_CHRDEV
#define MAX_CHRDEV	32
#endif
#ifndef MAX_BLKDEV
#define MAX_BLKDEV	32
#endif
#define MAX_DEV		MAX(MAX_CHRDEV, MAX_BLKDEV)

struct info
{
    unsigned long uptime;
    long m_to, m_us, m_fr, m_sh, m_bu, m_ca;
    long s_to, s_us, s_fr;
    unsigned long cpu_user, cpu_nice, cpu_sys, cpu_idle;
    unsigned long disk[5];
    unsigned long disk_r[5];
    unsigned long disk_w[5];
    unsigned long pgin, pgout, swin, swout;
    unsigned long *intr;	/* Phew. That's better. */
    unsigned long old_intr;
    unsigned long ctxt;
    unsigned long syscalls;
};

/* Prototypes. */

void window_init (void);
void tstp (int i);
void cont (int i);
void winsz (int i);
void quit (int i);
void set_echo (int i);

void fatal (const char *s,...);
void *my_xcalloc (size_t n, size_t s);
void init_terminal_data (void);
char *my_tgets (char *te);

char *make_version (FILE *verionfp);
FILE *myfopen (char *name);
char *hms (unsigned long t);
char *perc (unsigned long i, unsigned long t, int cpus);

#endif /* _PROCINFO_H */

/*
   Local variables:
   rm-trailing-spaces: t
   End:
 */


routines.c
/*

  routines.c

  routines for *info

  Date:        1995-04-15 23:51:55
  Last Change: 2001-02-24 23:43:21

  Copyright (C) 1995-2001 Sander van Malssen <svm@kozmix.cistron.nl>

  This software is released under the GNU Public Licence. See the file
  `COPYING' for details. Since you're probably running Linux I'm sure
  your hard disk is already infested with copies of this file, but if
  not, mail me and I'll send you one.

*/

static char *rcsid = "$Id: routines.c,v 1.24 2001/02/24 23:30:35 svm Exp svm $";

#include <errno.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <termcap.h>
#include <time.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/param.h>	/* for HZ -- should be in <time.h> ? */

#ifdef DEBUG
#define DMALLOC_FUNC_CHECK
#include <dmalloc.h>
#endif

#include "procinfo.h"

#ifndef NSIG
#ifdef _NSIG
#define NSIG _NSIG
#else
#define NSIG 32
#endif
#endif

extern char *cd, *ce, *cl, *cm, *ho, *se, *so, *ve, *vi;
extern int co, li, sg;
extern int fs, redrawn;
extern int nr_cpus;
extern FILE *versionfp;
extern char *version;

static struct termios oldstate, newstate;


/**** SIGNAL and SCREEN HANDLING ****/

void
window_init (void)
{
    struct sigaction sa;
    int i;

    tcgetattr (0, &oldstate);
    newstate = oldstate;
    newstate.c_lflag &= ~ICANON;
    newstate.c_lflag &= ~ECHO;
    tcsetattr (0, TCSANOW, &newstate);

    sa.sa_flags = 0;
    sigfillset (&sa.sa_mask);

    sa.sa_handler = quit;
    for (i = 1; i < NSIG; i++)
	sigaction (i, &sa, NULL);

    sa.sa_handler = winsz;
    sigaction (SIGWINCH, &sa, NULL);

    sa.sa_handler = cont;
    sigaction (SIGCONT, &sa, NULL);

    sa.sa_handler = tstp;
    sigaction (SIGTSTP, &sa, NULL);
}

void
tstp (int i)
{
    tcsetattr (0, TCSANOW, &oldstate);
    printf ("%s%s%s", ve, se, tgoto (cm, 0, li - 1));
    fflush (stdout);
    raise (SIGSTOP);
}

void
cont (int i)
{
    printf ("%s%s", cl, vi);
    tcsetattr (0, TCSANOW, &newstate);
    fflush (stdout);
}


/* This function originally stolen from top(1) (kmem-version). */

void
winsz (int i)
{
    struct winsize ws;

    co = li = 0;
    if (ioctl (1, TIOCGWINSZ, &ws) >= 0) {
	co = ws.ws_col;
	li = ws.ws_row;
    }
    if (isatty(fileno(stdout))) {
	if (co == 0)
	    co = tgetnum ("co");
	if (li == 0)
	    li = tgetnum ("li");
    }

    if (co == 0)
	co = 80;
    if (li == 0)
	li = 24;

    li -= 2;

    version = make_version (versionfp);

    redrawn = 1;
}

void
quit (int i)
{
    tcsetattr (0, TCSANOW, &oldstate);

    if (i == 0) {		/* This is an official exit. */
	printf ("%s%s%s", ve, se, tgoto (cm, 0, li + 1));
	exit (0);
    } else {
	printf ("%s%s%s", ve, se, tgoto (cm, 0, li));
	printf ("[%s]\n", sys_siglist[i]);
	exit (128 + i);
    }
}

void
set_echo (int i)
{
    if (i) {
	tcsetattr (0, TCSANOW, &oldstate);
	printf ("%s", ve);
	fflush (stdout);
    } else {
	printf ("%s", vi);
	tcsetattr (0, TCSANOW, &newstate);
    }
}

/*
  The next couple of functions are either directly stolen or else
  adapted from the GNU termcap manual, and therefore presumed
  copyrighted by the FSF.
  See the file `COPYING' for licencing details.
*/

void
fatal (const char *s,...)
{
    va_list l;

    va_start (l, s);
    vfprintf (stderr, s, l);
    va_end (l);

    exit (EXIT_FAILURE);
}

void *
my_xcalloc (size_t n, size_t s)
{
    void *p;

    p = calloc (n, s);
    if (p == 0)
	fatal ("my_xmalloc: memory exhausted\n");
    return p;
}

#ifdef __GNUC__
#define term_buffer 0
#else
static char term_buffer[2048];
#endif

void
init_terminal_data (void)
{
    char *termtype;
    int success;

    if (!(termtype = getenv ("TERM")))
	fatal ("Please specify a terminal type.\n");

    success = tgetent (term_buffer, termtype);
    if (success < 0)
	fatal ("Could not access the termcap data base.\n");
    if (success == 0)
	fatal ("Terminal type `%s' is not defined.\n", termtype);
}

char *
my_tgets (char *te)
{
#ifdef __GNUC__
#define BUFFADDR 0
#else
    /* Here we assume that an explicit term_buffer was provided to tgetent. */
    char *buffer = (char *) malloc (strlen (term_buffer));
#define BUFFADDR &buffer
#endif
    char *temp;

    /* Extract information we will use. */
    if ((temp = tgetstr (te, BUFFADDR)) != NULL)
	return temp;
    else
	return "";
}

/**** END OF FSF STUFF ****/

/**** MISCELLANEOUS ****/

char *
make_version (FILE * versionfp)
{
    char line[1024], myname[65];
    static char wheee[1024];
    char *p = line, *here, *there;
    size_t len;
    int ret; /* for gdb */

    /* These are the bits of /proc/version */
    char ver[64], host[1024], gcc[1024], date[1024], cpus[16];
    char compno[64];

    sprintf (cpus, "%dCPU", nr_cpus);

    gethostname (myname, 65);
    if (strchr(myname, '.') == NULL) {
	len = strlen(myname);
	if (len - 2 < 65) {		/* otherwise don't bother */
	    myname[len] = '.';
	    getdomainname(&myname[len+1], 65-2-len);
	}
    }

    fseek (versionfp, 0L, SEEK_SET);
    fgets (line, sizeof (line), versionfp);

    ret = sscanf (line, "Linux version %s (%[^)]) (gcc %[^)]) #%s %[^\n]",
		  ver, host, gcc, compno, date);

    if (ret != 5)	/* Damn egcs uses nested ()'s... */
	ret = sscanf (line, "Linux version %s (%[^)]) (gcc %[^(] (%*[^)])) #%s %[^\n]",
		      ver, host, gcc, compno, date);

    if (ret == 3) {	/* At least we've got ver & host right... */
	strcpy (gcc, "[can't parse]");
	strcpy (compno, "???");
	date[0] = 0;
    }

    /* BTW, from here we're free to re-use line[]. */

    here = strdup (myname);
    there = strdup (host);

    /*
      gcc[] may now be of the form
      	"version 2.7.2"
      for regular gcc releases or
      	"driver version 2.7.2p snapshot 970207 executing gcc version 2.7.2p"
      for snapshots. Or worse.
    */

    strcpy (line, gcc);

    if (strncmp (line, "version egcs", 12) == 0	||	/* This is egcs */
	(strncmp (line, "version pgcc", 12) == 0))	/* This is pgcc */
	ret = sscanf (line, "version %s", gcc);
    else if (line[0] == 'v')			/* This is "version 2.7.2" */
	strcpy (gcc, &line[8]);
    else if (line[0] == 'd') {		/* This is "driver version 2.7.2p" */
	char dee[64], dum[64];

	ret = sscanf (line, "driver version %s snapshot %s", dee, dum);
	if (ret == 2)
	    sprintf (gcc, "%s-%s", dee, dum);
	else
	    strcpy (gcc, "unknown");	/* This is "Or worse." */
    }

    /* First, let's see what happens if we put everything in. */
    sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s %s [%s]",
	     ver, host, gcc, compno, date, cpus, myname);

    /* Too long: truncate this system's name. */
    if ((len = strlen (wheee)) > (size_t) co) {
	for (p = myname; *p; p++)
	    if (*p == '.') {
		*p = '\0';
		break;
	    }
	sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s %s [%s]",
		 ver, host, gcc, compno, date, cpus, myname);
    }

    /* Too long: truncate compiling system's name. */
    if ((len = strlen (wheee)) > (size_t) co) {
	for (p = host; *p; p++)
	    if (*p == '.') {
		*p = '\0';
		break;
	    }
	sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s %s [%s]",
		 ver, host, gcc, compno, date, cpus, myname);
    }

    /* Restore hostnames, try again without date. */
    strcpy (myname, here);
    free (here);
    strcpy (host, there);
    free (there);

    /* Too long: try without date. */
    if ((len = strlen (wheee)) > (size_t) co)
	sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s [%s]",
		 ver, host, gcc, compno, cpus, myname);

    /* Too long: truncate this system's name. */
    if ((len = strlen (wheee)) > (size_t) co) {
	for (p = myname; *p; p++)
	    if (*p == '.') {
		*p = '\0';
		break;
	    }
	sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s [%s]",
		 ver, host, gcc, compno, cpus, myname);
    }

    /* Too long: truncate compiling system's name. */
    if ((len = strlen (wheee)) > (size_t) co) {
	for (p = host; *p; p++)
	    if (*p == '.') {
		*p = '\0';
		break;
	    }
	sprintf (wheee, "Linux %s (%s) (gcc %s) #%s %s [%s]",
		 ver, host, gcc, compno, cpus, myname);
    }

    return wheee;
}

FILE *
myfopen (char *name)
{
    FILE *fp;

    if ((fp = fopen (name, "r")) == NULL) {
	fprintf (stdout, "can't open file %s: %s\n", name, strerror (errno));
	exit (1);
    }
    return fp;
}

/* Note: we're using a static char array, so use this only once per printf. */
char *
hms (unsigned long t)
{
    unsigned int d, h, m, s;
    static char buf[22];

    t = t * 100 / HZ;
    d = (int) (t / 8640000);
    t = t - (long) (d * 8640000);
    h = (int) (t / 360000);
    t = t - (long) (h * 360000);
    m = (int) (t / 6000);
    t = t - (long) (m * 6000);
    s = (int) (t / 100);
    t = t - (long) (s * 100);
    if (d > 0)
	sprintf (buf, "%3ud %2u:%02u:%02u.%02u", d, h, m, s, (int) t);
    else
	sprintf (buf, "     %2u:%02u:%02u.%02u", h, m, s, (int) t);
    return buf;
}

/* Note: we're using a static char array, so use this only once per printf. */
char *
perc (unsigned long i, unsigned long t, int cpus)
{
    unsigned int v;
    static char buf[16];

    if ((signed long) i == -1 || t == 0)
	return "---.-%";

    v = (unsigned int) (i < 1000000 ?
			((1000 * i + t / 2) / t) :
			((i + t / 2000) / (t / 1000)));
    v /= cpus;

    /* if (v > 1000)
	return "+++.+%";
    else */
	sprintf (buf, "%3u.%u%%", v / 10, v % 10);
    return buf;
}


/*
   Local variables:
   rm-trailing-spaces: t
   End:
*/
